ts/screens/wallet/payment/ConfirmPaymentMethodScreen.tsx

```diff
@@ -204,14 +225,22 @@ class ConfirmToProceedTransactionScreen extends React.Component<Props, never> {
   }
 }
 
-/**
- * TODO: selectors will be reviewed in next pr
- *  @https://www.pivotaltracker.com/n/projects/2048617/stories/158395136
- */
-const mapStateToProps = (state: GlobalState): ReduxMappedStateProps => ({
-  card: selectedCreditCardSelector(state).getOrElse(UNKNOWN_CARD),
-  transaction: transactionForDetailsSelector(state).getOrElse(
-    UNKNOWN_TRANSACTION
-  )
+const mapStateToProps = (state: GlobalState): ReduxMappedStateProps => {
+  const card = selectedPaymentMethodSelector(state).getOrElse(UNKNOWN_CARD);
```

Comment:
is all this a valid state? (e.g. a transaction with unknown card, amount, fee) ?!?

Reply:
Well, if the payment process reached this point "normally", it means that a payment method must have been previously selected, so this should never fall back to `UNKNOWN_CARD`. 

In the unlikely event a user decides to tamper with the actions/state/anything else (I can see that happening if someone wants to contribute to the app development and starts changing things around) , this fallback option will prevent the UI from breaking too much (a bunch of question marks will pop up instead of showing a red screen).

Reply:
We can probably make a better job by encoding the fact the the payment method has been already selected in a type - we could have a union type like the following, acting like a FSM, encoding the valid types for each possible state:

```ts
type PaymentStep1 = {
  kind: "PaymentStep1";
  noticeNumber: string; 
}; 

type PaymentStep2 = {
  kind: "PaymentStep2";
  verifiedAmount: number;
}; 

type PaymentStep3 = {
  kind: "PaymentStep3";
  selectedPaymentMethod: Card;
};

type PaymentState = PaymentStep1 | PaymentStep2 | PaymentStep3;
```

Each screen will then have to check that the current state is the expected one (by checking the `kind` attribute).

In this case you avoid the "impossible" states like the one you mentioned (a payment method has been selected but you still have to check whether it's not empty).

Reply:
While this approach represents a more elegant solution, there are a couple of things I'd like to point out:
- Even though the "none" cases would no longer be handled in the `mapStateToProps` functions, the values are stored as `Option`al, so somewhere in the code there will be a `getOrElse` -- or similar -- behavior
- Providing the required data to each screen through a "PaymentStep" state essentially means that all the already-stored data needs to be replicated, thus doubling the data that needs to be updated any time an update is needed.

Reply:
Okay now I see your point & understand your proposed solution better. It would probably improve in readability, but I believe this still wouldn't entirely solve the original problem, i.e. the fact that some properties in the store could be undefined. 

Say the states available are:
```typescript
type PaymentStateManualDataEntry {
  kind: "PaymentStateManualDataEntry",
}
type PaymentStateSummary {
  kind: "PaymentStateSummary",
  rptId: RptId
}
...
type PaymentState = PaymentStateManualDataEntry | PaymentStateSummary | ... ;
```
Now, the state can be any of those possibilities. If I want to access `rptId`, from the `PaymentStateSummary`, I would need some sort of selector that, taking a `GlobalState` as input, returns an `RptId`. But since the payment state could be missing the `rptId` field, the selector would need to consider this scenario, and return an `Option<RptId>` (or `RptId | undefined` -- pretty much what happens [here](https://github.com/teamdigitale/italia-app/blob/7880fc1a673b1ca98a7a6b09e5ed695de4210de9/ts/store/reducers/authentication.ts#L98)). 

So, the result returned by the selector would need to be tested against these "empty" cases, and the original problem would present itself once more. Unless I misunderstood parts of the proposed approach?

Reply:
Each screen will expect some kind of `PaymentState`, the selector should check that the state matches one of the expected states or fail altogether (in this scenario we are also encoding in the component `Props` what are the expected states of this screen).

For instance:

```ts

type Props = {
  paymentState: PaymentStateSummary | undefined;
};

class PaymentSummaryScreen extends Component<Props> {

  render() {
    if(!this.props.paymentState) {
      return null;
    }
    return <View>...</View>;
  }
}

const mapStateToProps = (state: GlobalState): Props => ({
  paymentState: state.paymentState.kind === "PaymentStateSummary" ? state.paymentState : undefined
});

// ...
```

Reply:
validating the current state can/should happen also in the action reducers to validate that a certain action is expected only in a certain state and that only certain transitions are valid...

Reply:
Uh okay, that's pretty much what I started working on :D I just assumed you did not want to have to handle the "empty" case (`paymentState.kind !== "PaymentStateSummary"`) there, now anywhere else, and I could not figure out out to approach it differently.